# Lab 4: Data Wrangling [Total: 5 Points]

## Important
* Please ensure that you run the following two cells below before running any others. This will download all required files, as well as installing the necessary packages to ensure the code runs successfully. If you restart the kernel or your runtime session (in Colab), be sure to rerun this cell before running any others.
* This assignment recommends using **Google Colab**. If you are using **Anaconda Jupyter notebook/lab**, please ensure that this notebook is kept in a new folder. This is because the following commands will delete all files with the extensions .csv and .py before downloading the required files.

In [2]:
required_files = "https://github.com/mainuddin-rony/inst447-fall2024/raw/main/assignment/lab/lab4/required_files.zip"
! rm -rf tests
! rm -f required_files.zip *.csv *.py ._*.csv
! wget $required_files && unzip -j required_files.zip
! mkdir tests && mv *.py tests
! pip install otter-grader==5.5.0

--2024-10-07 15:20:03--  https://github.com/mainuddin-rony/inst447-fall2024/raw/main/assignment/lab/lab4/required_files.zip
Resolving github.com (github.com)... 140.82.114.4
Connecting to github.com (github.com)|140.82.114.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/mainuddin-rony/inst447-fall2024/main/assignment/lab/lab4/required_files.zip [following]
--2024-10-07 15:20:03--  https://raw.githubusercontent.com/mainuddin-rony/inst447-fall2024/main/assignment/lab/lab4/required_files.zip
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1275378 (1.2M) [application/zip]
Saving to: ‘required_files.zip’

required_files.zip  100%[===================>]   1.22M  --.-KB/s    in 0.06s   

2024-10-07 1

In [3]:
# Initialize Otter
import otter
grader = otter.Notebook()

## Instructions

The purpose of this assignment is for you to practice with common data wrangling operations, such as converting from long to wide format, merging, reshaping, and aggregating data frames. You will carry out this task in the present notebook, and use the notebook to document the various steps of the exercise and to answer all questions.

You will be working with a dataset describing orders made by customers. We have created a number of datasets for the purpose. A data dictionary listing all columns and their meaning is provided below. Follow the instructions in each question, and answer the accompanying questions. Note that some of the questions will require you to write a Python function to compute the correct answer.

<div class="alert alert-info">
This assignment will be graded automatically. To grade your submission we will run your code into the autograder. The autograder will compare the value returned by your function (i.e. the student's answer) against the correct answer. The autograder code itself is not included in this notebook, so double check that your code works correctly before submitting.
</div>

## Required skills

This lab will let you practice the following skills:
- Converting from long to wide
- Merging data frames
- Reshaping data frames
- Aggregating data frames

Before you start working on it, review Chapter 3 (p. 141‒157) of the textbook and/or the notebooks posted on ELMS under "Data Wrangling module".


## Data dictionary

This LAB includes multiple data files about an e-commerce website and grocery store, each corresponding to a different pandas table/data frame.

### File `orders.csv`

Each row in this table represents an order made by a particular customer. An order may contain multiple products.
- `order_id` &ndash; Unique order identifier;
- `user_id` &ndash; Unique user identifier;
- `order_number` &ndash; Order number (how many times a particular user has ordered using this service);
- `order_dow` &ndash; Day of week of the order (`0` &ndash; Monday &#8230; `6` &ndash; Sunday);
- `order_hour_of_day` &ndash; Hour of the day the order was placed;
- `days_since_prior_order` &ndash; Number of days since the user's previous order;

### File `products.csv`

Each row in this table represents a product in the store's inventory.
- `product_id` &ndash; Product ID (unique identifier for each product);
- `product_name` &ndash; Product name;
- `aisle_id` &ndash; Aisle ID (unique identifier for each aisle);
- `department_id` &ndash; Department ID (unique identifier for each department);

### File `order_products.csv`

Each row in this table information about a particular product that is part of an order.
- `order_id` &ndash; Unique order identifier;
- `product_id` &ndash; Product ID (Unique identifier for each product);
- `add_to_cart_order` &ndash; Order in which product was added to cart;
- `reordered` &ndash; Was the item reordered?

### File `OrganicAisle.csv`

Each row in this table contains information about a particular aisle. Only organic products are included in this table.
- `aisle_id` &ndash; Aisle ID (unique identifier for each aisle);
- `num_items` &ndash; Number of items ordered;
- `aisle` &ndash; Aisle name;

### File `ConventionalAisle.csv`
Each row in this table contains information about a particular aisle. Only conventional products are included in this table.
- `aisle_id` &ndash; Aisle ID (Unique identifier for each aisle);
- `num_items` &ndash; Number of items ordered;
- `aisle` &ndash; Aisle name.

## Q1

**Points**: 1

Write a function to called `busiest_dow` that takes a `filename` as a parameter and returns the day of the week with most orders.

Please make sure to provide the most suitable file as a parameter for this task. The function should then load the data into a data frame and return a Python integer (type `int`, not `int64`) representing a day of the week, with `0` for Monday and `6` for Sunday.

In [4]:
import pandas as pd

In [5]:
def busiest_dow(filename):
  df = pd.read_csv(filename)

  busy_day = df.groupby('order_dow').size().idxmax()
  return int(busy_day)

Use the cell below to run your function and see what it returns

In [6]:
import pandas as pd
filename = 'orders.csv' ## Mention the name of the file here
busiest_dow(filename)

0

This cell will autograde your answer

In [7]:
grader.check("q1")

q1 results: All test cases passed!

## Q2

**Points**: 1

Write a function called `items_per_order` that returns the number of items in each order.

Your function should not take any parameter. It should load the appropriate data file and it should return a pandas data frame, indexed by order ID, with the result stored in a column named `num_items`.

_**Hint**: this question requires to combine multiple datasets together before you can compute the aggregate column. Also, to rename a column after using groupby you can either use the `.rename()` method from the DataFrame class or the [Named Aggregation](https://pandas.pydata.org/docs/user_guide/groupby.html#named-aggregation) special syntax directly._

The resulting data frame should have exactly 5,000 rows. This is what the top 5 rows should look like:

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>num_items</th>
    </tr>
    <tr>
      <th>order_id</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>657</th>
      <td>19</td>
    </tr>
    <tr>
      <th>2203</th>
      <td>29</td>
    </tr>
    <tr>
      <th>2255</th>
      <td>12</td>
    </tr>
    <tr>
      <th>2654</th>
      <td>2</td>
    </tr>
    <tr>
      <th>2755</th>
      <td>4</td>
    </tr>
  </tbody>
</table>

and these are the bottom 5:

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>num_items</th>
    </tr>
    <tr>
      <th>order_id</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>3417262</th>
      <td>11</td>
    </tr>
    <tr>
      <th>3417499</th>
      <td>16</td>
    </tr>
    <tr>
      <th>3418430</th>
      <td>4</td>
    </tr>
    <tr>
      <th>3419803</th>
      <td>16</td>
    </tr>
    <tr>
      <th>3420412</th>
      <td>4</td>
    </tr>
  </tbody>
</table>

In [8]:
import pandas as pd

def items_per_order():
  df = pd.read_csv('order_products.csv')

  df = df.groupby('order_id').size().rename('num_items').reset_index()
  df.set_index('order_id', inplace=True)
  return df

Use the cell below to run your function and see what it returns

In [9]:
items_per_order()

num_items
order_id           
657              19
2203             29
2255             12
2654              2
2755              4
...             ...
3417262          11
3417499          16
3418430           4
3419803          16
3420412           4

[5000 rows x 1 columns]

This cell will autograde your answer

In [10]:
grader.check("q2")

q2 results: All test cases passed!

## Q3

**Points**: 1

Write a function called `avg_orders_by_dow` that returns the average number of orders (rounded to two decimals) for each day of the week.

Your function should not take any parameter. It should load the appropriate file(s) for this task into a data frame and it should return a pandas data frame, indexed by the day of the week, and with a column named `avg_orders`.

_**Hint**:  This question builds upon the previous one. To store the number of items into the orders data frame, make sure that to use the [.set_index()](https://pandas.pydata.org/docs/user_guide/indexing.html#set-reset-index) method on the Order ID column, otherwise you will get NaN._

This is what the data frame should look like:

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>avg_orders</th>
    </tr>
    <tr>
      <th>order_dow</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>11.15</td>
    </tr>
    <tr>
      <th>1</th>
      <td>10.39</td>
    </tr>
    <tr>
      <th>2</th>
      <td>9.47</td>
    </tr>
    <tr>
      <th>3</th>
      <td>9.21</td>
    </tr>
    <tr>
      <th>4</th>
      <td>9.58</td>
    </tr>
    <tr>
      <th>5</th>
      <td>9.66</td>
    </tr>
    <tr>
      <th>6</th>
      <td>11.25</td>
    </tr>
  </tbody>
</table>

In [11]:
import pandas as pd

def avg_orders_by_dow():
  orders = pd.read_csv('orders.csv')
  order_products = pd.read_csv('order_products.csv')

  order_item_count = order_products.groupby('order_id').size()
  orders.set_index('order_id', inplace=True)
  orders['number_of_items'] = order_item_count

  avg_orders = orders.groupby('order_dow')['number_of_items'].mean().round(2).rename('avg_orders').reset_index()
  avg_orders.set_index('order_dow', inplace=True)

  return avg_orders

Use the cell below to run your function and see what it returns

In [12]:
avg_orders_by_dow()

avg_orders
order_dow            
0               11.15
1               10.39
2                9.47
3                9.21
4                9.58
5                9.66
6               11.25

This cell will autograde your answer

In [13]:
grader.check("q3")

q3 results: All test cases passed!

## Q4

**Points**: 1

Write a function to select only organic aisles that have exactly 1 item only.

Your function should not take any parameter. It should load the appropriate file for this task into a data frame and it should return a pandas data frame.

This is what the data frame should look like:
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>aisle_id</th>
      <th>num_items</th>
      <th>aisle</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>40</th>
      <td>44</td>
      <td>1</td>
      <td>eye ear care</td>
    </tr>
    <tr>
      <th>77</th>
      <td>87</td>
      <td>1</td>
      <td>more household</td>
    </tr>
    <tr>
      <th>113</th>
      <td>124</td>
      <td>1</td>
      <td>spirits</td>
    </tr>
  </tbody>
</table>


In [14]:
import pandas as pd

def single_item_organic():
  df = pd.read_csv('OrganicAisle.csv')

  df = df[df['num_items'] == 1]
  return df

Use the cell below to run your function and see what it returns

In [15]:
single_item_organic()

aisle_id  num_items           aisle
40         44          1    eye ear care
77         87          1  more household
113       124          1         spirits

This cell will autograde your answer

In [16]:
grader.check("q4")

q4 results: All test cases passed!

## Q5

**Points**: 1

We want to know how many aisles have both organic and conventional items.

Write a function called `aisles_both` that combines the `organic` and `conventional` data sets into a single data frame.

Your function should not take any parameter. It should load the appropriate file(s) for this task into a data frame and it should return a pandas data frame.

_**Hint**: You may find it useful to use Pandas `merge` function. Since both data frames have common column names, make sure to specify the suffixes for any column with a common name, and to drop any redundant column._

Your data frame should have exactly 123 rows. This is what the first 5 rows should look like:
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>aisle_id</th>
      <th>num_items_org</th>
      <th>num_items_conv</th>
      <th>aisle</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>1</td>
      <td>895</td>
      <td>2679</td>
      <td>prepared soups salads</td>
    </tr>
    <tr>
      <th>1</th>
      <td>2</td>
      <td>36</td>
      <td>3801</td>
      <td>specialty cheeses</td>
    </tr>
    <tr>
      <th>2</th>
      <td>3</td>
      <td>2289</td>
      <td>15134</td>
      <td>energy granola bars</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4</td>
      <td>1320</td>
      <td>8539</td>
      <td>instant foods</td>
    </tr>
    <tr>
      <th>4</th>
      <td>5</td>
      <td>191</td>
      <td>2623</td>
      <td>marinades meat preparation</td>
    </tr>
  </tbody>
</table>

and the bottom 5 rows:
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>aisle_id</th>
      <th>num_items_org</th>
      <th>num_items_conv</th>
      <th>aisle</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>118</th>
      <td>129</td>
      <td>496</td>
      <td>8228</td>
      <td>frozen appetizers sides</td>
    </tr>
    <tr>
      <th>119</th>
      <td>130</td>
      <td>1528</td>
      <td>4848</td>
      <td>hot cereal pancake mixes</td>
    </tr>
    <tr>
      <th>120</th>
      <td>131</td>
      <td>2613</td>
      <td>8498</td>
      <td>dry pasta</td>
    </tr>
    <tr>
      <th>121</th>
      <td>132</td>
      <td>2</td>
      <td>244</td>
      <td>beauty</td>
    </tr>
    <tr>
      <th>122</th>
      <td>133</td>
      <td>2</td>
      <td>895</td>
      <td>muscles joints pain relief</td>
    </tr>
  </tbody>
</table>

In [18]:
def aisles_both():
 organic = pd.read_csv('OrganicAisle.csv')
 conventional = pd.read_csv('ConventionalAisle.csv')

 df = pd.merge(organic, conventional, on=['aisle_id', 'aisle'], suffixes=('_org', '_conv'))
 df = df[['aisle_id', 'num_items_org', 'num_items_conv', 'aisle']]

 return df

Use the cell below to run your function and see what it returns

In [19]:
aisles_both()

aisle_id  num_items_org  num_items_conv                       aisle
0           1            895            2679       prepared soups salads
1           2             36            3801           specialty cheeses
2           3           2289           15134         energy granola bars
3           4           1320            8539               instant foods
4           5            191            2623  marinades meat preparation
..        ...            ...             ...                         ...
118       129            496            8228     frozen appetizers sides
119       130           1528            4848    hot cereal pancake mixes
120       131           2613            8498                   dry pasta
121       132              2             244                      beauty
122       133              2             895  muscles joints pain relief

[123 rows x 4 columns]

This cell will autograde your answer

In [20]:
grader.check("q5")

q5 results: All test cases passed!

## Submission

Don't forget to run all cells in your notebook and then save it. To save, click on `File`, then select `Save/Save Notebook`. After that, download the notebook by going to `File`, then `Download` (for Anaconda notebook), and choosing `Download .ipynb` (for Colab). Finally, submit the notebook on Gradescope using the link found on ELMS.